src/view/renderer.js

```diff
@@ -494,52 +643,78 @@ export default class Renderer {
 			this._addInlineFiller( domDocument, expectedDomChildren, filler.offset );
 		}
 
-		const actions = diff( actualDomChildren, expectedDomChildren, sameNodes );
+		return {
+			actions: diff( actualDomChildren, expectedDomChildren, sameNodes ),
+			domElement,
+			actualDomChildren,
+			expectedDomChildren
+		};
+	}
 
-		let i = 0;
-		const nodesToUnbind = new Set();
+	/**
+	 * Finds DOM nodes which were replaced with the similar nodes (same tag name) in the view. All nodes are compared
+	 * within one `insert`/`delete` action group, for example:
+	 *
+	 * 		Actual DOM:		<p><b>Foo</b>Bar<i>Baz</i><b>Bax</b></p>
+	 * 		Expected DOM:	<p>Bar<b>123</b><i>Baz</i><b>456</b></p>
+	 * 		Input actions:	[ insert, insert, delete, delete, equal, insert, delete ]
+	 * 		Output actions:	[ insert, replace, delete, equal, replace ]
+	 *
+	 * @private
+	 * @param {Array} actions Actions array which is result of {@link module:utils/diff~diff} function.
+	 * @param {Array} actualDom Actual DOM children
+	 * @param {Array} expectedDom Expected DOM children.
+	 * @returns {Array} Actions array modified with `replace` actions.
+	 */
+	_findReplaceActions( actions, actualDom, expectedDom ) {
+		// If there is no both 'insert' and 'delete' actions, no need to check for replaced elements.
+		if ( actions.indexOf( 'insert' ) === -1 || actions.indexOf( 'delete' ) === -1 ) {
+			return actions;
+		}
+
+		function areSimilar( domNode1, domNode2 ) {
```

Comment:
These helper fns should not be defined in this scope. Every execution of this method creates them.